### PR TITLE
Preparing Stream for Alternate Database Drivers

### PR DESCRIPTION
**Pull request by faishal**
- Enable alternate drivers in `WP_Stream\Plugin::__construct()`, Replaced the hard-coded `WP_Stream\DB` reference with a value that can be filtered, defaulting to the WordPress Database.
- Created `WP_Stream\DB_Driver` interface, which help to implement Driver classes
- Created `WP_Stream\DB_Driver_WPDB` class and implemented `WP_Stream\DB_Driver` interface, Moved all WordPress database related operation and query into driver class.
- Updated `WP_Stream\DB` class to use driver passed in constructor. Class contains wrapper function to communicate with driver, Moved all query related function into database driver file.
- Updated `WP_Stream\List_Table` to use `WP_Stream\DB` wrapper method instead of `WP_Stream\Query` class methods.
- Some minnor changes in other files to adopt new DB class methods.
- Added related test cases.

### CODE REVIEW

**Comment by faishal** [comment]
@lukecarbis 

As we talked earlier on this, This branch is forked from `vip` branch. We have fixed couple of issues found during building an alternative driver and have added related test cases.

Also we tested this with `Stream to Papertrail` add-on to make sure we are not breaking anything.

**Comment by tlovett1** [comment]
Hey XWP, any update on this?

**Comment by lukecarbis** [comment]
Will test it today, but can you fix the unit test that's failing?

**Comment by lukecarbis** [comment]
cc. @faishal @tlovett1

File: classes/class-db-driver-wpdb.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		$prefix = apply_filters( 'wp_stream_db_tables_prefix', $wpdb->base_prefix );
+
+		$this->table         = $prefix . 'stream';
+		$this->table_meta    = $prefix . 'stream_meta';
+
+		$wpdb->stream        = $this->table;
+		$wpdb->streammeta    = $this->table_meta;
+
+		// Hack for get_metadata
+		$wpdb->recordmeta    = $this->table_meta;
```
Comment (lukecarbis) [inline classes/class-db-driver-wpdb.php]: Too much spacing after each of these properties. They can be aligned with less spaces.

File: classes/class-db-driver.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	/**
+	 * Returns array of existing values for requested column.
+	 * Used to fill search filters with only used items, instead of all items.
+	 *
+	 * @param string $column
+	 *
+	 * @return array
+	 */
+	public function get_column_values( $column );
+
```
Comment (lukecarbis) [inline classes/class-db-driver.php]: Extra line here can be removed.

File: classes/class-db.php [hunk 3]
```diff
@@ -101,37 +52,34 @@ public function insert( $record ) {
...
-		global $wpdb;
-
 		$fields = array( 'object_id', 'site_id', 'blog_id', 'user_id', 'user_role', 'created', 'summary', 'ip', 'connector', 'context', 'action' );
 		$data   = array_intersect_key( $record, array_flip( $fields ) );
 
-		$result = $wpdb->insert( $this->table, $data );
+		$meta = array();
+		foreach ( (array) $record['meta'] as $key => $vals ) {
+			// If associative array, serialize it, otherwise loop on its members
+			$vals = (is_array( $vals ) && 0 !== key( $vals )) ? array( $vals ) : $vals;
```
Comment (lukecarbis) [inline classes/class-db.php]: I think we need some extra spacing around the brackets here:

```
$vals = ( is_array( $vals ) && 0 !== key( $vals ) ) ? array( $vals ) : $vals;
```

File: classes/class-plugin.php [hunk 4]
```diff
@@ -73,14 +73,24 @@ public function __construct() {
...
 		if ( ! $this->db ) {
+			$error = 'Stream: Could not load chosen DB driver.';
+		} elseif ( ! $driver instanceof DB_Driver ) {
+			$error = 'Stream: DB driver must implement DB Driver interface.';
+		}
+
+		if ( $error ) {
 			wp_die(
-				esc_html__( 'Stream: Could not load chosen DB driver.', 'stream' ),
+				esc_html__( $error, 'stream' ),
```
Comment (lukecarbis) [inline classes/class-plugin.php]: Can't use a variable as the first arg of `esc_html__()`.

Instead:
```
		if ( ! $this->db ) {
			$error = __( 'Stream: Could not load chosen DB driver.', 'stream' );
		} elseif ( ! $driver instanceof DB_Driver ) {
			$error = __( 'Stream: DB driver must implement DB Driver interface.', 'stream' );
		}
```

and then just:

```
				esc_html( $error ),
```

File: classes/class-plugin.php [hunk 5]
```diff
@@ -97,6 +107,9 @@ public function __construct() {
 		// Add frontend indicator
 		add_action( 'wp_head', array( $this, 'frontend_indicator' ) );
 
+		// Change DB driver after plugin loaded if any add-on want to replace
+		add_action( 'plugins_loaded', array( $this, 'plugins_loaded' ) );
```
Comment (lukecarbis) [inline classes/class-plugin.php]: Good thinking.

File: classes/class-plugin.php [hunk 6]
```diff
@@ -208,4 +221,17 @@ private function locate_plugin() {
...
 		return self::VERSION;
 	}
+
+	/**
+	 * Change plugin database driver in case driver plugin loaded after stream
+	 */
+	public function plugins_loaded() {
+		// Load DB helper interface/class
+		$driver_class = apply_filters( 'wp_stream_db_driver', '\WP_Stream\DB_Driver_WPDB' );
+		$driver       = null;
```
Comment (lukecarbis) [inline classes/class-plugin.php]: Don't think we need this line.

File: classes/class-query.php [hunk 7]
```diff
@@ -305,20 +227,19 @@ public function query( $args ) {
...
 
+		$result = array();
 		/**
 		 * QUERY THE DATABASE FOR RESULTS
 		 */
-		$results = $wpdb->get_results( $query ); // @codingStandardsIgnoreLine $query already prepared
-
-		// Hold the number of records found
-		$this->found_records = absint( $wpdb->get_var( 'SELECT FOUND_ROWS()' ) );
+		$result['items'] = $wpdb->get_results( $query ); // @codingStandardsIgnoreLine $query already prepared
```
Comment (lukecarbis) [inline classes/class-query.php]: The query isn't prepared - also, it could be filtered.

Comment (lukecarbis) [inline classes/class-query.php]: We should change the query, where it's defined above, to:
```
		$query = "SELECT SQL_CALC_FOUND_ROWS %1$s
		FROM %2$s
		%3$s
		WHERE 1=1 %4$s
		%5$s
		%6$s";
```

And then, prepare properly on this line.

I realise that this isn't introduced by you, but it seems like an easy enough fix to include.

File: tests/tests/test-class-db-driver-wpdb.php [hunk 8]
```diff
@@ -0,0 +1,122 @@
+<?php
+namespace WP_Stream;
+
+class Test_DB_Driver_WPDB extends WP_StreamTestCase {
```
Comment (lukecarbis) [inline tests/tests/test-class-db-driver-wpdb.php]: Should the Query class be merged into this class?

File: classes/class-db-driver-wpdb.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
+<?php
+namespace WP_Stream;
+
+class DB_Driver_WPDB implements DB_Driver {
```
Comment (lukecarbis) [inline classes/class-db-driver-wpdb.php]: Should the Query class be merged into this class?

**Comment by lukecarbis** [comment]
@faishal This is looking really strong. I've made a whole bunch of comments, including some really small spacing changes (sorry about that).

There's three major things that need to happen before this can be merged:

1. We should have a discussion about the `Query` class, and whether it should be renamed, or merged with `DB_Driver_WPDB `
2. There are some `TODO` methods in `DB_Driver_WPDB` which still need implementation (`purge_storage`, `setup_storage`)
3. Some unit tests are failing

**Comment by dkotter** [comment]
This PR can be closed out in favor of #881.

**Pull request closed without merging**